add safe eval for browser and `eval` option

80avin opened the pull request:
Adding a safe alternative to Function/eval in browser.
This is also required since the chrome extensions & chrome apps have already become strict about safety.

I was trying to use it in a chrome extension but couldn't because of unsafe-eval. Hence this fix tries to introduce a safe evaluation method which satisfies Content Security Policy.

`eval: "safe"` is effective only in browser and doesn't uses a scripting engine which supports  a minimal form of javascript (no functions or complex expressions).

## PR description

<!-- Add the description of your PR here -->
* Add safe version of eval/Function for Content Security Policy
* Update test/index.html to use compiled module
* update demo html to include prefilled sample & prevent reload on submit
* babel/rollup related changes
* Add new option `eval: "safe" | "native" | boolean | (code, context) => value | Script)`
* Add new option `ignoreEvalError: boolean` which can be used to consider the errors in eval expressions as match.

## Checklist

- [x] - Added tests
- [x] - Ran `npm test`, ensuring linting passes
- [x] - Adjust README documentation if relevant

Fixes https://github.com/JSONPath-Plus/JSONPath/issues/60 , https://github.com/JSONPath-Plus/JSONPath/issues/182 , https://github.com/JSONPath-Plus/JSONPath/issues/169

80avin commented:
@brettz9  Could you please check and see if the approach is right ?

I particularly need advice on the rollup/babel/eslint configuration changes as I'm totally new to them.

Also, if possible, how should we go with the support of `"use strict"` and funciton syntax in scripts ?

If I get a green flag in these, I'll work on documentation.

brettz9 commented:
I don't have time to look at this now, and no guarantees about getting to it, but I think an optimal solution, if you can't get the multi-line evaluation supported, would be to allow the user to supply their own evaluator, e.g., optionally passing in `eval()`, so that users could still maintain support without too much added effort yet without requiring `eval` statements which create conflicts with content service policies such as Chrome extensions apparently enforce.

80avin commented:
user still has the choice to go with unsafe way using a new options `evalType` which can have values `none` or `native` or `safe`.
Thanks for the quick reply though. I'll move ahead with the cleanup & documentation. I should also be able to get multiline support.

80avin commented:
No test is failing under safe mode after I added the support of multiple expressions & assignment in latest change.
I will work on documentation & separate tests for safe/unsafe eval now.

80avin commented:
@brettz9 Looks like the PR is complete.
I added the tests, types, updated Readme and all tests are passing now.

I'm in no rush, so whenever you (or some other contributor) find time, please look into it and suggest changes.

brettz9 commented on .eslintignore:
```diff
@@ -1,6 +1,7 @@
-node_modules
-dist
+node_modules/
+dist/
```
Was this necessary?

brettz9 replied:
If not necessary, that's fine; I'm just curious if some environment behaves differently or what not and necessitated this.

80avin replied:
Not sure. probably some extension did this.
I'll revert.

brettz9 commented on .eslintrc.cjs:
```diff
@@ -17,6 +17,7 @@ module.exports = {
             'XMLHttpRequest'
         ]
     },
+    parser: '@babel/eslint-parser',
```
What language feature necessitated this?

80avin replied:
eslint gives this otherwise

```
JSONPath/rollup.config.js
  2:27  error  Parse errors in imported module '@rollup/plugin-node-resolve': Unexpected token { (25:11)  import/no-deprecated
  2:27  error  Parse errors in imported module '@rollup/plugin-node-resolve': Unexpected token { (25:11)  import/namespace

✖ 2 problems (2 errors, 0 warnings)
```

Not sure if there's some other better way to fix it or not.

brettz9 replied:
Instead of `parser`, try adding `parserOptions: {ecmaVersion: 2022}` and see if that resolves, or failing that, `parserOptions: {ecmaVersion: 2022, sourceType: 'module'}`.

80avin replied:
That worked.
Thanks !

brettz9 reviewed (changes requested):
I plan to give a deeper review later, but wanted to first address these smaller questions, and also ask that we maintain 100% testing coverage. (I see that the coverage badge is no longer reporting 100%, so some of the new lines of code apparently aren't covered in tests.)

I'd prefer we avoid the Babel ESLint parser unless there is really a compelling reason, though it'd be fine to increase the `parserOptions.ecmaVersion` to use the latest standard features in source.

From what I can see, the rest all looks very good.

jacobroschen commented:
> I don't have time to look at this now, and no guarantees about getting to it, but I think an optimal solution, if you can't get the multi-line evaluation supported, would be to allow the user to supply their own evaluator, e.g., optionally passing in `eval()`, so that users could still maintain support without too much added effort yet without requiring `eval` statements which create conflicts with content service policies such as Chrome extensions apparently enforce.

@brettz9, my two unsolicited cents are in agreement with your thoughts about this PR for a few couple reasons.
1. This adds dependencies for every consumer, even if they don't use the "safe" eval option.
2. While this PR passes all tests, `jsep` does not support all functionality that  `eval()` supports. This not only results in a breaking change, but will result in a third set of behavior that this library needs to support (default eval allow, `preventEval: true`, and now whatever `jsep` supports.
4. Adding a new config parameter such as `customEval` Instead of `evalType` would negate these downsides while providing some benefits benefits.
	1. Prevents the burden of a third mode for the maintenance of this library
	2. Provides greater flexibility for consumers (ie. a more limited eval function, or they want to use something like [QuickJS](https://github.com/justjake/quickjs-emscripten) for a safer, but just as powerful eval function)
	3. Removes the extra dependency for consumers that don't use the `evalType: 'safe'`
	4. Removes the weird ability to disable eval with two different config options `evalType` and `preventEval`
If we'd like to go down this path, there are two different config option styles that seem reasonable.
```typescript
const customEval = (code: string, context: any): any => {
    // evaluate code against context, return any matching result(s)
};
const jsonPath = new JSONPath({ eval: customEval });
jsonPath.evaulate('$.test');
```

An alternative API that fits into the current architecture a bit better, which would require the consumer to pass in a class to the config.
```typescript
class CustomEval {
  constructor(code: string) {
    // any compiling you need to do.
    this.code = code;
  }
  
  eval(context: any) {
    // however you'd like to evaluate the code (`this.code`) against `context`
    // return undefined, or match(es)
    return undefined;
  }
}

const jsonPath = new JSONPath({ eval: CustomEval });
jsonPath.evaluate('$.test');
```

In both scenarios, this library would still replace the shorthand properties(ex `@parentProperty`, `@property`, etc.) into `_$_parentProperty`, `_$_property`, etc while providing these values as context properties.

While option 2 has a bit more boilerplate, I think it provides more flexibility for consumers of this library and it enables automatic caching of the compiled script.

brettz9 commented:
@jacobroschen : Thank you for your thoughts. They are most welcome. However, I have a few thoughts in response:

As far as adding dependencies, the bundle size is rather good, no less considering what it is doing:

- 3.4 KB minified and gzipped: https://bundlephobia.com/package/jsep@1.3.8
- 661 B minfiied and gzipped: https://bundlephobia.com/package/@jsep-plugin/assignment@1.2.1
- 504 B minfiied and gzipped: https://bundlephobia.com/package/@jsep-plugin/regex@1.0.3

And although it does not support everything eval-wise, for the much more common use case of just accessing properties, it really does deliver something important.

I agree we can drop `preventEval` (as part of a breaking change), but that can be done as part of this PR as well. While this technically removes some functionality, I don't know how much this is going to harm anyone since it should not compromise security nor would I expect performance to significantly suffer by allowing what is already normally allowed in a jsonpath engine.

It has been a long-time goal for us to have a safe default, and recently we had still another request for a safe default: https://github.com/JSONPath-Plus/JSONPath/issues/182 . I rejected that because I didn't want to gut the core functionality our users have become accustomed to, but only because we didn't have an alternative.

I remain open to further thoughts, as well as to input of others, but unless there are concerns with jsep specifically, it seems like a hearty compromise without coming at much cost.

80avin commented:
I agree with @brettz9 . The primary reason why I decided to use jsep in this library was to provide a safe & lightweight evaluation out of the box. Another similar library [jsonpath](https://www.npmjs.com/package/jsonpath) achieves this by using [static-eval](https://npmjs.com/package/static-eval) and [esprima](https://npmjs.com/package/esprima) which have minified + gzip sizes as
* 19.4kB https://bundlephobia.com/package/static-eval@2.1.0
* 29kB https://bundlephobia.com/package/esprima@4.0.1

However, the idea of allowing the library user to provide his own `eval` alternative is also good. I'll update the PR with it.

80avin commented:
I have implemented the new API eliminating `preventEval` and `evalType` in favour of `eval`.
Please comment your suggestions.

guoliang commented:
will this PR ever be merged?

80avin commented:
@guoliang Thanks for reminding

@brettz9 I have removed the conflicts which had appeared during the time.
To me, the PR looks complete. Can you review on your part also ?

Also, since your last review, I added a new option `ignoreEvalError` which can be used to suppress the errors in the eval expressions.
The usecase is that if our expression passes only on certain types but gives errors on other types ( .toLowerCase() will error on anything other than string ), in that case we would want to consider the error as no match instead of crash.

brettz9 commented on test/index.html:
```diff
@@ -33,6 +33,7 @@ <h1>JSONPath Tests</h1>
         import './test.errors.js';
         import './test.escaping.js';
         import './test.eval.js';
+        import './test.safe-eval.js';
```
Not related to this line, but wanted to mention I'd like to preserve 100% code coverage and thereby avoid committing a badge with less than 100% coverage. If we really need to ignore some lines of code, we can do so with ignore statements.

80avin replied:
The code I added has 100% coverage already.

The only file which has no coverage is [bin/jsonpath-cli.js](https://github.com/JSONPath-Plus/JSONPath/blob/main/bin/jsonpath-cli.js).

Do you want me to handle that as well in this PR ?

brettz9 replied:
Ah, I see. It would be nice to restore 100%, sure, but I leave it up to you then.

80avin replied:
Added the tests for `bin/jsonpath-cli.js` as well.

I noticed that the cli is very minimal and lacks the options or good error reporting. But that should be out of scope for this PR.

---
PS:
I tried to be quick at deleting my inaccurate comment hoping no-one sees it, but I guess you also replied at the same moment.

For readers, the deleted message was like
> My code is already 100% test covered. The only file missing test coverage is [bin/jsonpath-cli.js](https://github.com/JSONPath-Plus/JSONPath/blob/324b8e52f7e12b321f07a625ebbde9340e692120/bin/jsonpath-cli.js) added in https://github.com/JSONPath-Plus/JSONPath/pull/206

I was inaccurate because the `ignoreEvalError` option I added at the last-minute had no tests and was breaking existing tests.

brettz9 commented on src/jsonpath.js:
```diff
@@ -1,4 +1,4 @@
-/* eslint-disable camelcase, unicorn/prefer-string-replace-all,
+/* eslint-disable camelcase, jsdoc/valid-types, unicorn/prefer-string-replace-all,
```
Please keep the rule enabled, and only disable on a case-by-case basis if really needed. If there is an issue with it, you may just need to add `.js` to the import statements?

80avin replied:
Done

brettz9 commented on src/jsonpath.js:
```diff
@@ -143,7 +162,8 @@ function JSONPath (opts, expr, obj, callback, otherTypeCallback) {
     this.wrap = hasOwnProp.call(opts, 'wrap') ? opts.wrap : true;
     this.sandbox = opts.sandbox || {};
-    this.preventEval = opts.preventEval || false;
+    this.eval = opts.eval === undefined ? 'safe' : opts.eval;
```
Especially given the importance of this option, I think we should throw if the argument is not one of the expected types (e.g., if there was a misspelling).

80avin replied:
Added the validation in
https://github.com/JSONPath-Plus/JSONPath/blob/d34dcce307f9116ca680845e1bcbf1433d51de66/src/jsonpath.js#L673-L675

Also a test
https://github.com/JSONPath-Plus/JSONPath/blob/d34dcce307f9116ca680845e1bcbf1433d51de66/test/test.eval.js#L29-L36

brettz9 reviewed (changes requested):
If you can make the mentioned changes, I think I'll be ready for a final review. Looks good so far...

80avin commented on .eslintignore:
```diff
@@ -4,3 +4,4 @@ docs/ts
 coverage
 ignore
 !*.js
+!.config.js
```
Done

80avin commented:
@brettz9 All resolved.

brettz9 commented:
I added one commit to fully pass linting and the testing scripts.

Did you mean for the `ignoreEvalError` option to be `ignoreEvalErrors` instead?

Also, please prepare a message for those upgrading to this breaking change version (e.g., what users of `preventEval` should now do instead).

80avin commented:
`ignorEvalErrors` sounds good as it'll be ignoring many errors, one for each path.

I hope this message suffices.

```
- Breaking change: remove `preventEval` property. Prefer `eval: false` instead.
- Breaking change: changed behaviour of `eval` property. In the browser, `eval`/`Function` won't be used by default to evaluate expressions. Instead, we'll safely evaluate using a subset of JavaScript. 
- feat: add `eval` property.
- feat: add `ignoreEvalErrors` property.
```

brettz9 commented:
If you want to take one last look, I think it should be ready for a release now.

80avin commented:
Tried few expressions in browser demo only, and it looks good to me.

brettz9 commented:
Released as v9.0.0. Thanks for your contributions!
